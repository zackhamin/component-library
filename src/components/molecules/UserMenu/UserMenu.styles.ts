import { colors } from 'Styles';
import styled from 'styled-components';

export const StyledTrigger = styled.button`
	display: flex;
	border: 0;
	align-items: center;
	justify-content: center;
	background: ${colors.emerald5};
	padding: 1.2rem;
	cursor: pointer;
	border-radius: 0.8rem;
	transition: 0.2s;

	&:hover,
	&:focus {
		background: ${colors.emerald10};
	}
`;

export const StyledList = styled.ul`
	display: flex;
	flex-direction: column;
	list-style: none;
	padding: 0;
	margin-bottom: 1.4rem;
	border-bottom: 1px solid ${colors.grey500};

	li {
		padding: 0.8rem 0;
		font-size: 1.4rem;

		&.disabled {
			pointer-events: none;
		}
	}
`;
