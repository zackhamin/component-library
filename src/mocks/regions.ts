import { Region } from '@Types/region';

export const mockRegions: Region[] = [
	{
		name: 'United Kingdom',
		countryCode: 'GB',
		locale: 'en-GB',
		siteId: '10',
	},
	{
		name: 'Ireland',
		countryCode: 'IE',
		locale: 'en-IE',
		siteId: '30',
	},
	{
		name: 'The Netherlands',
		countryCode: 'NL',
		locale: 'nl-NL',
		siteId: '40',
	},
	{
		name: 'Belgium',
		countryCode: 'BE',
		locale: 'nl-BE',
		siteId: '60',
	},
];
