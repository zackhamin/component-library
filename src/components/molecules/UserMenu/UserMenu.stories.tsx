import { colors } from 'Styles';
import { mockRegions } from 'MockData/regions';
import { Notification } from 'Atoms/Notification';
import { UserMenu, UserMenuProps } from './UserMenu';

export default {
	title: 'molecules/UserMenu',
	component: UserMenu,
};

export const UserMenuStory = (args: UserMenuProps) => (
	<>
		<Notification
			backgroundColor={colors.grey200}
			hasBorder
			color={colors.pomegranate}
			margin="0 0 1.6rem"
		>
			This component was implemented for Aurora and might not work as intended
			when used in other apps.
		</Notification>
		<div style={{ display: 'flex', justifyContent: 'flex-end' }}>
			<UserMenu {...args} />
		</div>
		<p>
			This component can be used together with the <code>useLocalStorage</code>{' '}
			hook. we export from the <code>hooks</code> directory.
		</p>
		<h2>Example usage</h2>
		<pre style={{ fontSize: '1.6rem', margin: 0 }}>
			{`import { useLocalStorage } from 'Hooks/useLocalStorage';
import { UserMenu } from 'Molecules/UserMenu';

const Layout = () => {
	const [selectedRegion, setSelectedRegion] = useLocalStorage('region', '10');

	return (
		<div>
			<UserMenu
				onSignOut={signOut}
				regions={regions}
				selectedRegion={selectedRegion}
				setSelectedRegion={setSelectedRegion}
			/>
		</div>
	);
};

export { Layout };`}
		</pre>
	</>
);

UserMenuStory.storyName = 'UserMenu';
UserMenuStory.args = {
	regions: mockRegions,
	selectedRegion: '10',
};
UserMenuStory.argTypes = {
	onSignOut: { action: 'signed out' },
	setSelectedRegion: { action: 'selected region' },
};
