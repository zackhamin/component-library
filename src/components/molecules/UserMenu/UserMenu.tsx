import { Button } from 'Atoms/Button';
import { Heading } from 'Atoms/Heading';
import { Popover } from 'Atoms/Popover';
import { User } from 'Atoms/Icons';

import { colors } from 'Styles';
import { StyledList, StyledTrigger } from './UserMenu.styles';

import { Region } from '@Types/region';

export interface UserMenuProps {
	onSignOut: () => void;
	regions: Region[];
	selectedRegion: Region['siteId'];
	setSelectedRegion: (region: string) => void;
	disableRegionSelection?: boolean;
}

const UserMenu = ({
	onSignOut,
	regions,
	selectedRegion,
	setSelectedRegion,
	disableRegionSelection = false,
}: UserMenuProps) => {
	return (
		<Popover
			id="user"
			title="User Settings"
			align="right"
			trigger={
				<StyledTrigger className="user-icon">
					<User color={colors.emerald} fontSize="2.2rem" />
				</StyledTrigger>
			}
		>
			<Heading level={3} size="1.6rem">
				Regions
			</Heading>
			<StyledList>
				<li className="disabled">
					<strong>
						{regions.find(({ siteId }) => siteId === selectedRegion)?.name}
					</strong>
				</li>
				{regions
					.filter(({ siteId }) => siteId !== selectedRegion)
					.map(({ name, siteId }) => (
						<li key={siteId}>
							<Button
								onClick={() => setSelectedRegion(siteId)}
								disabled={disableRegionSelection}
								style="text"
								size="small"
							>
								{name}
							</Button>
						</li>
					))}
			</StyledList>
			<Button size="small" style="text" onClick={onSignOut}>
				Sign out
			</Button>
		</Popover>
	);
};

export { UserMenu };
