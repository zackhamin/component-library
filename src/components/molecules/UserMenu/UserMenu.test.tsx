import { mockRegions } from 'MockData/regions';
import { UserMenu } from '../UserMenu';
import { fireEvent, render } from '@testing-library/react';

describe('UserMenu', () => {
	it('renders correctly', () => {
		const { container } = render(
			<UserMenu
				onSignOut={jest.fn}
				setSelectedRegion={jest.fn}
				regions={mockRegions}
				selectedRegion="10"
			/>
		);
		expect(container.firstChild).toMatchSnapshot();
	});

	it('disables region buttons if disabled prop is passed', () => {
		const { getByText, getByRole } = render(
			<UserMenu
				onSignOut={jest.fn}
				setSelectedRegion={jest.fn}
				regions={mockRegions}
				selectedRegion="10"
				disableRegionSelection
			/>
		);

		fireEvent.click(getByRole('button'));
		const regionButton = getByText('Belgium');

		expect(regionButton).toBeDisabled();
	});
});
