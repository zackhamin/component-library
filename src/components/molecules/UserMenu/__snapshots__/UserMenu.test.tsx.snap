// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserMenu renders correctly 1`] = `
.c0 {
  position: relative;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
}

.c1 {
  display: inline;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  border: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  background: #F2F9F8;
  padding: 1.2rem;
  cursor: pointer;
  border-radius: 0.8rem;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c2:hover,
.c2:focus {
  background: #E6F3F2;
}

<div
  class="c0"
>
  <span
    class="c1"
  >
    <button
      aria-controls="user"
      aria-expanded="false"
      class="c2 user-icon"
      id="user-trigger"
    >
      <svg
        aria-hidden="true"
        color="#00594C"
        fill="none"
        focusable="false"
        font-size="2.2rem"
        height="1em"
        viewBox="0 0 22 26"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M13.357 15.357H8.643A8.643 8.643 0 0 0 0 24c0 .868.704 1.571 1.571 1.571H20.43c.867 0 1.571-.703 1.571-1.571a8.643 8.643 0 0 0-8.643-8.643ZM2.406 23.214a6.292 6.292 0 0 1 6.237-5.5h4.714c3.2 0 5.849 2.404 6.237 5.5H2.406ZM11 13A6.286 6.286 0 1 0 11 .428 6.286 6.286 0 0 0 11 13Zm0-10.214a3.933 3.933 0 0 1 3.929 3.928A3.933 3.933 0 0 1 11 10.643a3.934 3.934 0 0 1-3.929-3.929A3.933 3.933 0 0 1 11 2.786Z"
          fill="currentColor"
        />
      </svg>
    </button>
  </span>
</div>
`;
